app/models/backend_api_config.rb

```diff
@@ -43,4 +45,13 @@ def with_subpath?
   def path=(value)
     super(ConfigPath.new(value).path)
   end
+
+  private
+
+  def destroy_usage_limits
+    usage_limits = UsageLimit
+                     .of_metric(backend_api_metrics.pluck(:id))
+                     .of_plan(service.application_plans.pluck(:id))
+    usage_limits.destroy_all unless usage_limits.empty?
+  end
```

Comment:
Shouldn't we create a `through` association and have `dependent: :destroy`? Also are usage limits in the background deletion tree of `BackendAPI`?

Reply:
maybe two `through` associations, e.g. `metric_usage_limits` and `plan_usage_limits` if that makes sense

Reply:
I am not sure which association, and `through` what :smiling_face_with_tear: 

> Also are usage limits in the background deletion tree of `BackendAPI`?

They are not, because usage limits belong to `Plan`, so they are in background deletion tree of `Plan`.

Reply:
e.g.
```
  has_many :plan_usage_limits, through: :plan, dependent: :destroy
```

Reply:
> e.g.
> 
> ```
>   has_many :plan_usage_limits, through: :plan, dependent: :destroy
> ```

On the `BackendApiConfig` model? (aka "backend usage"). I am not sure...

There would need to be double hop then: `BackendApiConfig` > `Service` > `Plan` > `UsageLimit`, but also you will only going to be interested (and destroy if needed) in those usage limits that are related to metrics that belong to `BackendApiConfig`'s backend... 
I don't know, I think this scope will be too complicated to handle. Honestly, I wouldn't do it, given the current `before_destroy` already does the trick.

Reply:
Ok, I couldn't find official information but in a [blog](https://medium.com/@zmommaerts/how-to-access-deeply-nested-associations-in-ruby-on-rails-using-activerecord-and-serializers-38c9d448efee) I read that deeply nested associations were not recommended. So callback should be fine in such case.

Reply:
In fact I read no such discouragements in upstream documentation. See "Nested Associations" here:

https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html

In my opinion using an association would be cleaner than a custom method. Why use custom logic when we can delegate to rails inner logic?